File: assignments/haskell/atbash-cipher/atbash-cipher_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import Atbash (encode)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
       [ TestList atbashTests ]

atbashTests :: [Test]
atbashTests = map TestCase
  [ "ml" @=? encode "no"
  , "bvh" @=? encode "yes"
  , "lnt" @=? encode "OMG"
  , "lnt" @=? encode "O M G"
  , "nrmwy oldrm tob" @=? encode "mindblowingly"
  , "gvhgr mt123 gvhgr mt" @=? encode "Testing, 1 2 3, testing."
  , "gifgs rhurx grlm" @=? encode "Truth is fiction."
  , "gsvjf rxpyi ldmul cqfnk hlevi gsvoz abwlt" @=?
    encode "The quick brown fox jumps over the lazy dog."
  ]

File: assignments/go/atbash-cipher/atbash_cipher_test.go
package atbash

import "testing"

var tests = []struct {
	expected string
	s        string
}{
	{"ml", "no"},
	{"ml", "no"},
	{"bvh", "yes"},
	{"lnt", "OMG"},
	{"lnt", "O M G"},
	{"nrmwy oldrm tob", "mindblowingly"},
	{"gvhgr mt123 gvhgr mt", "Testing, 1 2 3, testing."},
	{"gifgs rhurx grlm", "Truth is fiction."},
	{"gsvjf rxpyi ldmul cqfnk hlevi gsvoz abwlt", "The quick brown fox jumps over the lazy dog."},
}

func TestAtbash(t *testing.T) {
	for _, test := range tests {
		actual := Atbash(test.s)
		if actual != test.expected {
			t.Errorf("Atbash(%s): expected %s, actual %s", test.s, test.expected, actual)
		}
	}
}

func BenchmarkAtbash(b *testing.B) {
	b.StopTimer()
	for _, test := range tests {
		b.StartTimer()

		for i := 0; i < b.N; i++ {
			Atbash(test.s)
		}

		b.StopTimer()
	}
}
